packages/lib/src/vis/tiles/TiledHeatmap.tsx

```diff
@@ -0,0 +1,68 @@
+import { clamp, range } from 'lodash';
+
+import { getInterpolator } from '../heatmap/utils';
+import { useFrameRendering } from '../hooks';
+import { useAxisSystemContext } from '../shared/AxisSystemContext';
+import TiledHeatmapLayer from './TiledHeatmapLayer';
+import type { TilesApi } from './api';
+import { useDataPerPixel } from './hooks';
+import type { ColorMapAttributes } from './models';
+
+interface Props extends ColorMapAttributes {
+  api: TilesApi;
+  displayLowerResolutions?: boolean;
+  qualityFactor?: number;
+}
+
+function TiledHeatmap(props: Props) {
+  const {
+    api,
+    displayLowerResolutions = true,
+    qualityFactor = 1, // 0: Lower quality, less fetch; 1: Best quality
+    ...colorMapAttrs
+  } = props;
+
+  const { visSize } = useAxisSystemContext();
+  const { xDataPerPixel, yDataPerPixel } = useDataPerPixel();
+
+  const nLevels = api.levelsSize.length;
+
+  const quality = 1 - clamp(qualityFactor, 0, 1);
+  const levelOfDetail = Math.min(
+    Math.floor(Math.log2(Math.max(1, xDataPerPixel, yDataPerPixel)) + quality),
+    Math.max(nLevels - 1, 0)
+  );
+
+  // displayLowerResolutions selects which levels of detail layers are displayed:
+  // true: lower resolution layers displayed behind the current one
+  // false: only current level of detail layer is displayed
+  const lods = displayLowerResolutions
+    ? range(levelOfDetail, nLevels).reverse()
+    : [levelOfDetail];
```

Comment:
As discussed, I wonder if storing the layer sizes in the opposite order in the first place would simplify a few things. Conceptually, I find that when I'm zoomed out, the level of detail should be 0 and should increase as I zoom in. It would then make even more sense to keep the `imageSize` out of `layerSizes`, since it would no longer be the first size in the array but the last.

Reply:
I inverted the order for storing layer sizes, but I kept all sizes in the same array and added a getter for the base one.